AMD SOF based generic SoundWire machine driver

vijendarmukunda opened the pull request:
This patch series includes below listed changes.
1) Move and rename Common helper functions in Intel machine driver to common place holder which can be made generic.
2) Move and rename Codec helper functions which can be made generic.
3) Move the common Codec info list which is generic
4) Add pci revision id field in mach_params for platform distinguish 
5) AMD SOF based SoundWire machine generic driver.

This series is rebase version of https://github.com/thesofproject/linux/pull/4955 .

vijendarmukunda committed:
ASoC: amd: acp: add soundwire machines for acp6.3 based platform

Add Soundwire machines for acp6.3 based platform.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: SOF: amd: add alternate machines for acp6.3 based platform

Add SoundWire machines as alternate machines for acp6.3 based platform.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: rename ignore_pch_dmic variable name

Rename 'ignore_pch_dmic' variable name as 'ignore_internal_dmic' to make
it more generic across different platforms to represent internal dmic.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire dai type macros

Move SoundWire dai type macros to common header file(soc_sdw_util.h) and
rename the macros to make it generic. It will be used by other platform
machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: soc-acpi: add pci revision id field in mach params structure

Few IP's may have same PCI vendor id and device id and
different revision id. Add revision id field to the
'snd_soc_acpi_mach_params' so that using this field platform
specific implementation can be added in machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: SOF: amd: update mach params subsystem_rev variable

Add pci_rev variable in acp sof driver private data structure and assign
this value to mach_params structure subsystem_rev variable.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move dai id common macros

Move dai id common macros from intel SoundWire generic driver to
soc_sdw_utils.h file so that it can be used by other platform machine
driver. Rename macros with 'ASOC' as prefix.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: amd: acp: add sof based soundwire generic machine driver

Add sof based Soundwire generic driver for amd platforms.
Currently support added for ACP6.3 based platforms.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: sdw_utils: rename amp codec quirk macros

To make it generic, rename amp codec quirk macros.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw-utils: move soundwire machine driver soc ops

Move Intel SoundWire generic machine driver soc ops to common place
so that it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire machine driver common structures

Move intel generic SoundWire machine driver common structures to
soc_sdw_utils.h file. These structures will be used in other platform
SoundWire machine driver code.
Rename macro 'SOF_SDW_MAX_DAI_NUM' as 'SOC_SDW_MAX_DAI_NUM'.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire machine driver helper functions

Move below Intel SoundWire machine driver helper functions to
soc_sdw_utils.c file so that it can be used by other platform machine
driver.
- is_unique_device()
- get_codec_name()
Rename the functions by adding 'asoc_sdw' tag as a prefix.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move dmic codec helper functions

Move generic dmic codec helper functions implementation to
sdw_utils folder so that these functions can be used by other platform
machine drivers.
Rename the helper function 'sof_sdw_dmic_init' as 'asoc_sdw_dmic_init'
as this helper function will be used by SOF and legacy(No DSP) stacks.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rtk dmic helper functions

Move rtk SoundWire dmic helper functions implementation to sdw_utils
folder to make it generic. Rename dmic helper functions by adding
'asoc_sdw' as prefix.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt712 sdca helper functions

Move RT712 SDCA codec helper file to sdw_utils folder.
Rename 'rt712_spk_rtd_init' function as 'asoc_sdw_rt712_spk_rtd_init'.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt722 sdca helper functions

Move RT722 SDCA codec helper file to sdw_utils folder to make it generic.
Rename helper function with 'asoc_sdw' tag.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: split soundwire machine driver private data

Split intel generic SoundWire machine driver private data into two
structures. One structure is generic one which can be used by other
platform machine driver and the other one is intel specific one.
Move generic machine driver private data to soc_sdw_utils.h.
Define a void pointer in generic machine driver private data structure
and assign the vendor specific structure in mc_probe() call.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt5682 codec helper functions

Move rt5682 sdw codec helper functions to common place holder to make it
generic. Rename 'rt5682_rtd_init' function as 'asoc_sdw_rt5682_rtd_init'.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rtk jack common helper functions

Move RTK codec jack common helper functions to common place holder
(sdw_utils folder) to make it generic so that it will be used by
other platform machine driver code.
Rename codec helper functions by adding 'asoc_sdw' tag.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt700 and rt711 codec helper functions

Move RT700 and RT711 Soundwire codec helper functions to common
place holder so that it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rtk amp codec helper functions

Move RTK amp codec helper functions related implementation to common
place holder to make it generic so that these helper functions will be
used by other platform machine driver modules.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move cirrus soundwire codec helper functions

To make it generic, move Cirrus Soundwire codec helper functions to
common place holder so that it can be used by other platform machine
driver. Rename the codec helper functions.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move maxim codec helper functions

Move maxim codec helper functions to common place holder so that
it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: rename codec_info and dai_info structures names

To make it generic, rename structure 'sof_sdw_codec_info' as
'soc_sdw_codec_info' and 'sof_sdw_dai_info' as 'soc_sdw_dai_info'.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire codec_info_list structure

SoundWire 'codec_info_list' structure is not a platform specific one.
Move codec_info_list structure to common file soc_sdw_utils.c.
Move codec helper functions which uses codec_info_list structure to common
place holder and rename the function by adding _sdw tag. This will allow
to use 'codec_info_list' structure and it's helper functions in other
platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move machine driver dai link helper functions

Move machine driver dai link helper functions to common place holder,
so that it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: sdw_utils: refactor sof_sdw_card_late_probe function

Refactor sof_sdw_card_late_probe() function and derive a generic
function soc_sdw_card_late_probe() function which can be used by
SoundWire generic machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -0,0 +1,23 @@
+int asoc_sdw_hw_free(struct snd_pcm_substream *substream);
+void asoc_sdw_shutdown(struct snd_pcm_substream *substream);
+
+#endif
```
maybe move to include/sound/ ?

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -10,6 +10,50 @@
+	unsigned long quirk;
+};
+
+struct sof_sdw_codec_info {
```
rename as soc_sdw_codec_info ?

vijendarmukunda replied:
already renamed in later patch.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -10,6 +10,50 @@
+
+struct sof_sdw_codec_info;
+
+struct sof_sdw_dai_info {
```
rename as sof_sdw_dai_info?

vijendarmukunda replied:
We have renamed in it later patch.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -10,6 +10,50 @@
 #include <sound/soc.h>
+#include <sound/soc-acpi.h>
+
+#define SOF_SDW_MAX_DAI_NUM             8
```
Rename macro 'SOF_SDW_MAX_DAI_NUM' as 'SOC_SDW_MAX_DAI_NUM'.

that seems to be missed?

plbossart commented on sound/soc/amd/ps/ps-sdw-mach.c:
```diff
@@ -0,0 +1,216 @@
+	.trigger = acp63_sdw_trigger,
+	.hw_free = acp63_sdw_hw_free,
+	.shutdown = acp63_sdw_shutdown,
+};
```
why can't you use the functions you added earlier, e.g.

````c
static const struct snd_soc_ops acp63_sdw_ops = {
	.startup = asoc_sdw_startup,
	.prepare = asoc_sdw_prepare,
	.trigger = asoc_sdw_trigger,
	.hw_params = asoc_sdw_hw_params,
	.hw_free = asoc_sdw_hw_free,
	.shutdown = asoc_sdw_shutdown,
};
````

vijendarmukunda replied:
It's my bad. It shouldn't be included in this patch.  This is a just test version machine driver file for PS platform. 
We already working on implementing legacy machine driver similar to sof based soundwire generic machine driver.
Will drop these changes

plbossart commented on sound/soc/amd/ps/ps-sdw-mach.c:
```diff
@@ -0,0 +1,216 @@
+	dev_dbg(&pdev->dev, "entry=%s\n", __func__);
+	card = &acp63_sdw_card;
+	card->dev = &pdev->dev;
+	usleep_range(200 * 1000, 250 * 1000);
```
oh that seems so suspicious right there....

That's a NAK from me, there shouldn't be any hard-coded sleep time in a probe, that's a clear sign of something fishy.

vijendarmukunda replied:
Sorry this is a test version file added by mistake. We are working on implementing Soundwire generic machine driver for legacy(NO DSP) stack in similar to SOF based Soundwire generic machine driver. This file will be dropped.

plbossart commented on sound/soc/amd/ps/ps-sdw-mach.c:
```diff
@@ -0,0 +1,216 @@
+MODULE_AUTHOR("Vijendar.Mukunda@amd.com");
+MODULE_DESCRIPTION("TAS2783 audio support");
+MODULE_LICENSE("GPL");
+MODULE_ALIAS("platform:" DRV_NAME);
```
and what does this have to do with "ASoC: intel/sdw_utils: move rtk jack common helper functions" anyways, that's a patch split or git squash that didn't go well.

vijendarmukunda replied:
Please ignore this file changes. Will drop this file.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -13,6 +13,8 @@
 #include <sound/soc-acpi.h>
 
 #define SOF_SDW_MAX_DAI_NUM             8
+#define MAX_NO_PROPS 2
```
SOC_SDW prefix

vijendarmukunda replied:
will fix it.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -42,23 +42,23 @@ struct sof_sdw_dai_info {
 	const bool ignore_internal_dmic;
 	const struct snd_soc_ops *ops;
-	struct sof_sdw_dai_info dais[SOF_SDW_MAX_DAI_NUM];
+	struct asoc_sdw_dai_info dais[SOF_SDW_MAX_DAI_NUM];
```
this rename should have been done earlier IMHO

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -42,23 +42,23 @@ struct sof_sdw_dai_info {
 	const bool ignore_internal_dmic;
 	const struct snd_soc_ops *ops;
-	struct sof_sdw_dai_info dais[SOF_SDW_MAX_DAI_NUM];
+	struct asoc_sdw_dai_info dais[SOF_SDW_MAX_DAI_NUM];
```
and you still have SOF_SDW in the define for the size.

vijendarmukunda replied:
In later patch we redefined the macro.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.h:
```diff
@@ -113,6 +116,15 @@ const char *asoc_sdw_get_codec_name(struct device *dev,
+							 int *dai_index);
+
+int asoc_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd);
+
```
Now that I think about it, it's probably easier to do the renames in one shot

a) rename everything in sof_sdw. It'll be massively different but that's just a rename.
b) move code in the utils stuff.
c) add new things.

That will reduce the scope for errors/omissions.

vijendarmukunda replied:
Agreed. will make the changes.

plbossart reviewed (changes requested):
I am having second thoughts on the incremental renames, which contains omissions.
It might be just easier to rename everything in sof_sdw, then do the code moves.

Also the PS stuff is badly broken, a sleep as the first step in a probe isn't going to fly.

vijendarmukunda commented:
Removed ps-sdw-mach.c file and rebased the patch series and posted V2 series.
Will post the V3 patch series after fixing the review comments and also will resolve the merge conflict in sof_sdw.c file.

vijendarmukunda commented:
Requires patches rebase to resolve the merge conflict and needs to redefine the patch series.will open the new PR with fresh patch series
